[clangd] Support .clangd command line modifications for C++ modules

petr-polezhaev opened the pull request:
Tunnels `Manger` object into the `ScanningAllProjectModules` so it can be used to perform necessary command-line modifications (which also adds `--resources` path previously added there explicitly). This allows using the experimental C++ modules support with gcc.

This was discussed in the issue with @ChuanqiXu9 and @kadircet

Closes #112635

petr-polezhaev committed:
[clangd] Support .clangd command line modifications in ScanningAllProjectModules

ChuanqiXu9 reviewed (commented):
Do we need to update `buildModuleFile` in `clang-tools-extra/clangd/ModulesBuilder.cpp` to update the commands to build the module interfaces?

petr-polezhaev commented:
> Do we need to update `buildModuleFile` in `clang-tools-extra/clangd/ModulesBuilder.cpp` to update the commands to build the module interfaces?

It uses `GlobalCompilationDatabase::getCompileCommand` (if I correctly understand what you mean) which is overridden by `OverlayCDB::getCompileCommand` which does call the `Mangler` before returning the command, so that should already be covered

ChuanqiXu9 reviewed (approved):
Thanks. This looks good to me. And please give @kadircet some time to give it a look. I'll try to land this in the next week if no more comments came in.

petr-polezhaev commented:
Ok. I'll wait for review confirmation and then fix the clang-format thing and merge new main

kadircet commented on clang-tools-extra/clangd/unittests/PrerequisiteModulesTest.cpp:
```diff
@@ -191,6 +211,29 @@ export module M;
+  EXPECT_TRUE(
+      ProjectModules->getRequiredModules(getFullPath("M.cppm")).empty());
+
+  CDB.addExcludedFlag("-invalid-unknown-flag");
```
instead of adding a new method/field to test fixture, i'd just call `ProjectModules.setCommandProvider` here.

kadircet reviewed (changes requested):
thanks for working on this!

petr-polezhaev commented on clang-tools-extra/clangd/ScanningProjectModules.cpp:
```diff
@@ -32,6 +32,9 @@ namespace {
 /// interfere with each other.
 class ModuleDependencyScanner {
 public:
+  using CommandProvider =
```
In my original approach I just made the original CommandMangler (a struct) shared_ptr and passed it here. It is also safer than to capture a reference since that's the only way how I can pass it inside the ProjectModules with the unique_function. I don't know why it's converted to a function object. Should I do it this way (shared_ptr) or keep the function thing?

It would be basically
```
virtual void setCommandMangler(std::shared_ptr<CommandMangler> Mangler);
```

kadircet replied:
Sorry I was just trying to say we don't need to have the same type-alias both in ModuleDependencyScanner and ProjectModules, you already unified them, so my comment is addressed.

petr-polezhaev commented on clang-tools-extra/clangd/ScanningProjectModules.cpp:
```diff
@@ -32,9 +32,6 @@ namespace {
 public:
-  using CommandProvider =
-      llvm::unique_function<void(tooling::CompileCommand &, PathRef) const>;
-
```
Or `using CommandMangler = ProjectModules::CommandMangler`?

petr-polezhaev commented on clang-tools-extra/clangd/GlobalCompilationDatabase.cpp:
```diff
@@ -833,8 +833,8 @@ bool OverlayCDB::setCompileCommand(PathRef File,
   auto MDB = DelegatingCDB::getProjectModules(File);
-  MDB->setCommandProvider([&Mangler = Mangler](tooling::CompileCommand &Command,
-                                               PathRef CommandPath) {
+  MDB->setCommandMangler([&Mangler = Mangler](tooling::CompileCommand &Command,
```
@kadircet this is what I was talking about with unsafe reference capturing. The solution (while keeping lambda compatibility for unit-tests) would be to make the function copyable but pass the Mangler by the `shared_ptr` into the very first instance.

kadircet commented on clang-tools-extra/clangd/ProjectModules.h:
```diff
@@ -9,12 +9,17 @@
 
 namespace clang {
+namespace tooling {
+class CompileCommand;
```
ah sorry I missed the dep on `tooling::CompileCommand`. feel free to re-add `#include "clang/Tooling/CompilationDatabase.h"` instead of the forward-decl.

kadircet commented on clang-tools-extra/clangd/unittests/PrerequisiteModulesTest.cpp:
```diff
@@ -38,10 +39,6 @@ class MockDirectoryCompilationDatabase : public MockCompilationDatabase {
 
   void addFile(llvm::StringRef Path, llvm::StringRef Contents);
 
-  std::unique_ptr<ProjectModules> getProjectModules(PathRef) const override {
-    return scanningProjectModules(MockedCDBPtr, TFS);
-  }
-
```
i think we still need to override this

petr-polezhaev replied:
True. My bad, removed that accidentally - should have ran unit-tests before updating.

kadircet reviewed (approved):
thanks!

petr-polezhaev commented:
rebased on the latest main squashing all fixes

ChuanqiXu9 commented:
Given @kadircet is approved and @petr-polezhaev is marked as `first-time contributor`, I'll merge this by assuming you don't have the commit access.

ChuanqiXu9 commented:
@petr-polezhaev please update the mail address. We don't like github.reply mail address.

petr-polezhaev commented:
Didn't knew that's a thing now. Removed the checkbox over the private email, should be ok now. Somehow missed the clang-format issue again (I thought I fixed it) - updated to calm it down

mikaelholmen commented on clang-tools-extra/clangd/GlobalCompilationDatabase.cpp:
```diff
@@ -830,6 +830,16 @@ bool OverlayCDB::setCompileCommand(PathRef File,
+                                              PathRef CommandPath) {
+    Mangler(Command, CommandPath);
+  });
+  return std::move(MDB);
```
Hi @petr-polezhaev 

I see that clang (18.1.8) complains on this move:
```
12:29:43 ../llvm-project/clang-tools-extra/clangd/GlobalCompilationDatabase.cpp:840:10: error: moving a local object in a return statement prevents copy elision [-Werror,-Wpessimizing-move]
12:29:43   840 |   return std::move(MDB);
12:29:43       |          ^
12:29:43 ../llvm-project/clang-tools-extra/clangd/GlobalCompilationDatabase.cpp:840:10: note: remove std::move call here
12:29:43   840 |   return std::move(MDB);
12:29:43       |          ^~~~~~~~~~   ~
12:29:43 1 error generated.
```

petr-polezhaev replied:
:disappointed: Can't do anything about it - I don't have write permissions. Can you revert it and I'll resubmit the review?

kadircet replied:
oops sorry for missing that, should be fixed with a2063ba7ffdbbb4faf5da5f32739ab761c2e4289

mikaelholmen replied:
Thanks!
